container/layer.bzl

```diff
@@ -0,0 +1,169 @@
+  layer = ctx.outputs.layer
+  build_layer_exec = ctx.executable.build_layer
+  args = [
+      "--output=" + layer.path,
+      "--directory=" + directory,
+      "--mode=" + ctx.attr.mode,
+  ]
+
+  args += ["--file=%s=%s" % (f.path, magic_path(ctx, f)) for f in files]
+  args += ["--file=%s=%s" % (f.path, path) for (path, f) in file_map.items()]
+  args += ["--empty_file=%s" % f for f in empty_files or []]
+  args += ["--tar=" + f.path for f in tars]
+  args += ["--deb=" + f.path for f in debs]
+  for k in symlinks:
+    if ":" in k:
+      fail("The source of a symlink cannot container ':', got: %s" % k)
+  args += ["--link=%s:%s" % (k, symlinks[k]) for k in symlinks]
+  arg_file = ctx.new_file(ctx.label.name + "-layer.args")
+  ctx.file_action(arg_file, "\n".join(args))
+  ctx.action(
+      executable = build_layer_exec,
+      arguments = ["--flagfile=" + arg_file.path],
+      inputs = files + file_map.values() + tars + debs + [arg_file],
+      outputs = [layer],
+      use_default_shell_env = True,
+      mnemonic="ImageLayer",
+  )
+  return layer, _sha256(ctx, layer)
+
+def zip_layer(ctx, layer):
+  zipped_layer = _gzip(ctx, layer)
+  return zipped_layer, _sha256(ctx, zipped_layer)
+
+LayerInfo = provider()
+
+def _impl(ctx, files=None, file_map=None, empty_files=None, directory=None,
+          symlinks=None, output=None, env=None, debs=None, tars=None):
+  """Implementation for the container_layer rule.
+
+  Args:
+    ctx: The bazel rule context
+    files: File list, overrides ctx.files.files
+    file_map: Dict[str, File], defaults to {}
+    empty_files: str list, overrides ctx.attr.empty_files
+    directory: str, overrides ctx.attr.directory
+    symlinks: str Dict, overrides ctx.attr.symlinks
+    env: str Dict, overrides ctx.attr.env
+    debs: File list, overrides ctx.files.debs
+    tars: File list, overrides ctx.files.tars
+"""
+  file_map = file_map or {}
+  files = files or ctx.files.files
+  empty_files = empty_files or ctx.attr.empty_files
+  directory = directory or ctx.attr.directory
+  symlinks = symlinks or ctx.attr.symlinks
+  env = env or ctx.attr.env
+  debs = debs or ctx.files.debs
+  tars = tars or ctx.files.tars
+  # Generate the unzipped filesystem layer, and its sha256 (aka diff_id)
+  unzipped_layer, diff_id = build_layer(ctx, files=files, file_map=file_map,
+                                        empty_files=empty_files,
+                                        directory=directory, symlinks=symlinks,
+                                        debs=debs, tars=tars)
+  # Generate the zipped filesystem layer, and its sha256 (aka blob sum)
+  zipped_layer, blob_sum = zip_layer(ctx, unzipped_layer)
+  # Returns constituent parts of the Container layer as provider
+  return [LayerInfo(zipped_layer=zipped_layer,
+                    blob_sum=blob_sum,
+                    unzipped_layer=unzipped_layer,
+                    diff_id=diff_id,
+                    env=env)]
+
+_layer_attrs = dict({
```

Comment:
I didn't handle this comments yet, since I am still not sure about how to change `container_image` to a macro. 

There are still some attrs which are `container_image` only, e.g. `legacy_run_behavior`, `labels` etc., which are not in `container_layer` rule.  Also how to deal with the backward compatibility of the existing usage of `container_image` rule? 

So how about 1) I re-use all the `container_layer` attrs and 2) re-use the `container_layer` impl for the related attrs, but still keep `container_image` as a rule?

Reply:
I'm ok with punting on this for now, but I'd still like the rules to head this general direction to avoid repeating logic.  This is the motivation for some of my comments like `_image_config`.

The idea is that `container_image` would still be a rule, but likely `_container_image`, and `container_image` becomes a macro that wraps `container_layer` and `_container_image`.